src/coreclr/jit/valuenum.cpp

```diff
@@ -7680,106 +8067,29 @@ void Compiler::fgValueNumberAssignment(GenTreeOp* tree)
-                    }
-                    else
-                    {
-                        // We do know the field sequence.
-                        // The "lclFld" node will be labeled with the SSA number of its "use" identity
-                        // (we looked in a side table above for its "def" identity).  Look up that value.
-                        ValueNumPair oldLhsVNPair =
-                            lvaTable[lclFld->GetLclNum()].GetPerSsaData(lclFld->GetSsaNum())->m_vnPair;
-                        newLhsVNPair = vnStore->VNPairApplySelectorsAssign(oldLhsVNPair, lclFld->GetFieldSeq(),
-                                                                           rhsVNPair, // Pre-value.
-                                                                           lclFld->TypeGet());
-                    }
-                }
-
-                lvaTable[lclFld->GetLclNum()].GetPerSsaData(lclDefSsaNum)->m_vnPair = newLhsVNPair;
+            GenTreeLclFld* lclFld      = lhs->AsLclFld();
+            unsigned       storeOffset = lclFld->GetLclOffs();
+            unsigned       storeSize   = lclFld->GetSize();
 
-                JITDUMP("Tree [%06u] assigned VN to local var V%02u/%d: ", dspTreeID(tree), lclFld->GetLclNum(),
-                        lclDefSsaNum);
-                JITDUMPEXEC(vnpPrint(newLhsVNPair, 1));
-                JITDUMP("\n");
-            }
-            else if (lvaVarAddrExposed(lclFld->GetLclNum()))
-            {
-                // This side-effects ByrefExposed.  Just use a new opaque VN.
-                // As with GT_LCL_VAR, we could probably use MapStore here and MapSelect at corresponding
-                // loads, but to do so would have to identify the subset of address-exposed locals
-                // whose fields can be disambiguated.
-                ValueNum heapVN = vnStore->VNForExpr(compCurBB, TYP_HEAP);
-                recordAddressExposedLocalStore(tree, heapVN DEBUGARG("local field assign"));
-            }
-            else
+            // TODO-PhysicalVN: with the physical VN scheme, we no longer need the field sequence checks.
+            if ((lclFld->GetFieldSeq() == FieldSeqStore::NotAField()) && ((lclFld->gtFlags & GTF_VAR_USEASG) != 0))
             {
-                JITDUMP("Tree [%06u] assigns to non-address-taken local var V%02u; excluded from SSA, so value not"
-                        "tracked.\n",
-                        dspTreeID(tree), lclFld->GetLclNum());
+                // Invalidate the whole local.
+                rhsVNPair.SetBoth(vnStore->VNForExpr(compCurBB, lvaGetDesc(lclFld)->TypeGet()));
+                storeOffset = 0;
+                storeSize   = lvaLclExactSize(lclFld->GetLclNum());
             }
+
+            // TODO-PhysicalVN: remove this quirk, it was added to avoid diffs.
+            bool normalize = ((lclFld->gtFlags & GTF_VAR_USEASG) != 0);
```

Comment:
What does the type otherwise end up being for these? Type of a physical map?

Reply:
Whatever the RHS's type was we will write directly into the descriptor and then read when numbering corresponding `LCL_VAR`s.

So you can have `TYP_DOUBLE` `LCL_VAR`s with `TYP_LONG` VNs and other combinations: it's why we have code in VN and global constant propagation compensating for these mismatches.

Reply:
Not sure I understand, that would not be marked as a partial definition, would it?
I'm confused about the cases where this normalization changes types for a partial definition. Wouldn't it imply the local was not properly normalized beforehand?

Reply:
For a true partial definition (i. e. `!LoadStoreIsEntire`), normalization to the type of the local will always be performed because for that case `GTF_VAR_USEASG` will always be set.

> Wouldn't it imply the local was not properly normalized beforehand?

Correct, it is the case that today we can and do have "denormalized" locals, because of this quirk (not performing normalization for full definitions) and because of some `BYREF <-> TYP_I_IMPL` case.

Reply:
I guess I see how it can happen if you first have a "wrongly" typed full store followed by a partial store. Is the quirk just there to match the existing cast inserted in `VNApplySelectorsAssignTypeCoerce`?. FWIW, I see no assertion failures over SPMI after applying this patch:
```diff
diff --git a/src/coreclr/jit/compiler.h b/src/coreclr/jit/compiler.h
index fee5d2356c2..933c35143f0 100644
--- a/src/coreclr/jit/compiler.h
+++ b/src/coreclr/jit/compiler.h
@@ -4697,7 +4697,7 @@ public:
                                  ssize_t              offset,
                                  unsigned             storeSize,
                                  ValueNumPair         value,
-                                 bool                 normalize = true);
+                                 bool                 normalize = true, bool partial = false);
 
     void fgValueNumberArrayElemLoad(GenTree* loadTree, VNFuncApp* addrFunc);
 
diff --git a/src/coreclr/jit/valuenum.cpp b/src/coreclr/jit/valuenum.cpp
index ef71522ff40..dc2cb2c6e69 100644
--- a/src/coreclr/jit/valuenum.cpp
+++ b/src/coreclr/jit/valuenum.cpp
@@ -4562,7 +4562,7 @@ void Compiler::fgValueNumberLocalStore(GenTree*             storeNode,
                                        ssize_t              offset,
                                        unsigned             storeSize,
                                        ValueNumPair         value,
-                                       bool                 normalize)
+                                       bool                 normalize, bool partial)
 {
     // Should not have been recorded as updating the GC heap.
     assert(!GetMemorySsaMap(GcHeap)->Lookup(storeNode));
@@ -4593,8 +4593,10 @@ void Compiler::fgValueNumberLocalStore(GenTree*             storeNode,
 
         if (normalize)
         {
+            var_types prevType = vnStore->TypeOfVN(newLclValue.GetLiberal());
             // We normalize types stored in local locations because things outside VN itself look at them.
             newLclValue = vnStore->VNPairForLoadStoreBitCast(newLclValue, varDsc->TypeGet(), lclSize);
+            assert(!partial || (prevType == vnStore->TypeOfVN(newLclValue.GetLiberal())));
             assert((genActualType(vnStore->TypeOfVN(newLclValue.GetLiberal())) == genActualType(varDsc)));
         }
 
@@ -8089,7 +8091,7 @@ void Compiler::fgValueNumberAssignment(GenTreeOp* tree)
 
             // TODO-PhysicalVN: remove this quirk, it was added to avoid diffs.
             bool normalize = ((lclFld->gtFlags & GTF_VAR_USEASG) != 0);
-            fgValueNumberLocalStore(tree, lclFld, storeOffset, storeSize, rhsVNPair, normalize);
+            fgValueNumberLocalStore(tree, lclFld, storeOffset, storeSize, rhsVNPair, normalize, true);
         }
         break;
```

Reply:
> I guess I see how it can happen if you first have a "wrongly" typed full store followed by a partial store.

Right.

> Is the quirk just there to match the existing cast inserted in `VNApplySelectorsAssignTypeCoerce`

The correct (-- not quirked) behavior is to always normalize, so the purpose of the quirk is to skip normalization for full definitions.

https://github.com/dotnet/runtime/blob/6111bc7fcfc5445e8d48b2321edb5ba2b1461ae5/src/coreclr/jit/valuenum.cpp#L7730-L7737

Of course, if we already have the correctly-typed map for a partial definition (as you show, the 99.99% case), the normalization is a no-op, so I suppose the question is if we should make the logic a bit like this:
```
if (LoadStoreIsEntire(...) { /* normalize /* } else { /* assume normalized */ }.
```
It would work; but first we have to make sure all definitions are normalizing.

Reply:
Ah ok, got it, the quirk is the opposite of what I thought :) Now it makes more sense.